### PR TITLE
Fixed #26034 -- Fixed incorrect index handling on PostgreSQL on Char/TextField with unique=True and db_index=True.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/26034

### CODE REVIEW

File: django/db/backends/postgresql/schema.py [hunk 1]
```diff
@@ -111,13 +111,13 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
             new_db_params, strict,
         )
         # Added an index? Create any PostgreSQL-specific indexes.
-        if ((not old_field.db_index and new_field.db_index) or (not old_field.unique and new_field.unique)):
+        if not old_field.db_index and not old_field.unique and (new_field.db_index or new_field.unique):
```
Comment (charettes) [inline django/db/backends/postgresql/schema.py]: I'm curious as to why we're even checking `unique`.

Does `not old_field.db_index and new_field.db_index` works?

Comment (charettes) [inline django/db/backends/postgresql/schema.py]: Oh are we creating the `_like` indexes even when `unique=True`? I wasn't expecting that.

File: tests/schema/tests.py [hunk 2]
```diff
@@ -1769,9 +1765,58 @@ def test_alter_field_add_index_to_textfield(self):
...
+
+        # Alter to add unique=True (should add 1 index)
+        old_field = BookWithoutAuthor._meta.get_field('title')
+        new_field = CharField(max_length=100, db_index=True, unique=True)
+        new_field.set_attributes_from_name('title')
+        with connection.schema_editor() as editor:
+            editor.alter_field(BookWithoutAuthor, old_field, new_field, strict=True)
+        indexes = self.get_constraints_for_column(BookWithoutAuthor, 'title')
+        # TODO: check if 3 indexes is correct
+        self.assertEqual(3, len(indexes))
```
Comment (charettes) [inline tests/schema/tests.py]: The fact a normal index and a unique one is tracked by [#24082](https://code.djangoproject.com/ticket/24082). The third one is the `_like` index.

File: tests/schema/tests.py [hunk 3]
```diff
@@ -1769,9 +1765,58 @@ def test_alter_field_add_index_to_textfield(self):
...
+        self.assertEqual(2, len(title_indexes))
+
+        # Alter to add db_index=True (should add 1 index)
+        old_field = Tag._meta.get_field('slug')
+        new_field = SlugField(db_index=True, unique=True)
+        new_field.set_attributes_from_name('slug')
+        with connection.schema_editor() as editor:
+            editor.alter_field(Tag, old_field, new_field, strict=True)
+        indexes = self.get_constraints_for_column(Tag, 'slug')
+        self.assertEqual(2, len(indexes))
```
Comment (charettes) [inline tests/schema/tests.py]: This is inconsistent but I think the patch can land as is and the test be modified later on based on the direction of [#24082](https://code.djangoproject.com/ticket/24082).

**Comment by charettes** [comment]
It would be great if you actually inspected the index details to make sure the correct ones are created.

File: django/db/backends/postgresql/schema.py (hunk 1 above)
Comment (timgraham) [inline django/db/backends/postgresql/schema.py]: Correct

**Comment by timgraham** [comment]
As you noted, I think there are other bugs here, but the one in the ticket should be fixed.

**Comment by charettes** [comment]
Looking at the index name should be good enough, LGTM.

**Pull request merged**